### PR TITLE
[Feature] Allow user to specify a fraction of the GPU memory. 

### PR DESCRIPTION
**Pull request by CalvinXKY**
Add a new function,  torch.cuda.set_per_process_memory_fraction(fraction, device), to torch.cuda.  Related:  #18626
The fraction (float type, from 0 to 1) is used to limit memory  of cashing allocator on GPU device .  One can set it on any visible GPU. The allowed memory equals total memory * fraction. It will raise an OOM error when  try to apply GPU memory more than the allowed value. This function is similar to Tensorflow's per_process_gpu_memory_fraction  
Note， this setting is just limit the cashing allocator in one process. If you are using multiprocess, you need to put this setting in to the subprocess to limit its GPU memory, because subprocess could have its own allocator.

## usage
In some cases, one needs to split a GPU device as two parts. Can set limitation before GPU memory using. 
Eg. device: 0, each part takes half memory, the code as follows: 
```
torch.cuda.set_per_process_memory_fraction(0.5, 0)
```
There is an example to show what it is.
```python
import torch
torch.cuda.set_per_process_memory_fraction(0.5, 0)
torch.cuda.empty_cache()
total_memory = torch.cuda.get_device_properties(0).total_memory
# less than 0.5 will be ok:
tmp_tensor = torch.empty(int(total_memory * 0.499), dtype=torch.int8, device='cuda')
del tmp_tensor
torch.cuda.empty_cache()
# this allocation will raise a OOM:
torch.empty(total_memory // 2, dtype=torch.int8, device='cuda')

"""
It raises an error as follows: 
RuntimeError: CUDA out of memory. Tried to allocate 5.59 GiB (GPU 0; 11.17 GiB total capacity; 0 bytes already allocated; 10.91 GiB free; 5.59 GiB allowed; 0 bytes reserved in total by PyTorch)
"""
```

### CODE REVIEW

File: torch/cuda/memory.py [hunk 1]
```diff
@@ -72,6 +72,33 @@ def caching_allocator_delete(mem_ptr):
     torch._C._cuda_cudaCachingAllocator_raw_delete(mem_ptr)
 
 
+def set_memory_fraction(fraction, device: Union[Device, int] = None) -> None:
+    r"""Set memory fraction for a device.
+    The fraction is used to limit allocated memory on a CUDA device.
```
Comment (VitalyFedyunin) [inline torch/cuda/memory.py:77]: The statement is incorrect as you are only limiting memory used for caching allocator, so running tow processes with fraction of 0.5 not going to be possible (but expected as per description) as some of the memory going to be used by operators and cuda context.

Comment (CalvinXKY) [reply torch/cuda/memory.py:77]: @VitalyFedyunin This does not work for multiprocess limiting, becasue all subproesses will get its own space  and do not share info to each other. However, in most cases, one can use it to split memory, if he/she knows put the setting in subprocess.  So should I just change the statement or do some more, such as to deal with the multiprocess situation?

Comment (CalvinXKY) [reply torch/cuda/memory.py]: @VitalyFedyunin hi, I've just changed the statment and the name of this function, to make it more clearly.

**Review by VitalyFedyunin — APPROVED** [review]

**Comment by CalvinXKY** [comment]
@VitalyFedyunin hi, I noticed the issue #58466 and tried to fix it. However, I could not find a way to solve it with only a little change. A way I currently found is to use nvml lib to get the process info and modify the allocator total memory data. Then, user's max memory would be limited on a GPU, but it could not work when in dock env. because, the host process PID not equals the container PID if one does not use PID mapping.
Through this way to satisfy part cases and only need to introduce "nvml.h" lib and change less than 100 lines code. Is this OK?
Another method is to change function name and clarify its using situation in doc. Any suggestions？

**Pull request closed without merging**